### PR TITLE
Extend parsing errors add gbif class names

### PR DESCRIPTION
**Pull request by damianooldoni**
This PR aims to extend parsing errors functionality to name related data (checklists). Such functionality is already present in current `rgbif` package for occurrence data. You can apply piping for filtering (in or out) occurrences with some issues and you can spread multiple issues (`split` option), expand issue (`expand` option) codes to more descriptive names or both at the same time (`split_expand` option). This means you can do the following:

```r
res <- occ_search(
      geometry = 'POLYGON((30.1 10.1, 10 20, 20 40, 40 40, 30.1 10.1))',
      limit = 50
)
# Parsing output by issue
res %>% occ_issues(gass84) # pipe operator included in rgbif package

### remove data rows with certain issue classes
res %>% occ_issues(-cudc, -cdreps)

### split issues into separate columns
res %>% occ_issues(mutate = "split")
res %>% occ_issues(-cudc, -mdatunl, -cdreps, mutate = "split")
res %>% occ_issues(gass84, mutate = "split")

### expand issues to more descriptive names
gg <- res %>% occ_issues(mutate = "expand")

### split and expand
res %>% occ_issues(mutate = "split_expand")

```

This PR extends such functionality to name related data (checklists):

```r
out <- name_usage(name = "Lupus")

# Parsing output by issue
out %>% name_issues(clasna)

### remove data rows with certain issue classes
out %>% name_issues(-bbmn, -clasna)

### split issues into separate columns
out %>% name_issues(mutate = "split")
out %>% name_issues(-scina, -clasna, mutate = "split")
out %>% name_issues(bbmn, mutate = "split")

### expand issues to more descriptive names
out %>% name_issues(mutate = "expand")

### split and expand
out %>% name_issues(mutate = "split_expand")
```

From the code point of view, this was not as straightforward as thought. So, I discussed with Scott Chamberlain the best way to do it, while maintaining the typical distinction between `occ_*` and `name_*` functions. See discussion at https://github.com/ropensci/rgbif/issues/326. Based on it, next changes have been applied:

1. the table storing the mapping of all possible issues, called `nameissues` and the function to get it, `name_issues()`, has been changed to `gbifissues()` and `gbif_issues()` respectively. In this way we can use `name_issues()` as the exact counterpart of `occ_issues()`. 
2. A column called `type` has been added to `gbif_issues()` in order to classify the issues: `type = occurrence` for occurrence related issues, `type = name` for name related issues. Again, discussed together with Scott in https://github.com/ropensci/rgbif/issues/326.
3. A call to `occ_issues()` or `name_issues()` will first get a look to the type of the issue: `occ_issues()` accepts only occurrence related issues and `name_issues()` accepts only name related issues. Informative errors are raised otherwise. This check runs in `occ_issues()` and `name_issues()` by calling the internal function `check_issues(type)`:  `check_issues(type = "occurrence", ...)` or `check_issues(type = "name", ...)` respectively.
4. as function `occ_issues()` accepts only data of class `gbif` (from `occ_search()`) or `gbif_data` (from `occ_data()`), Scott and I agreed to provide class `gbif` to name data as well (this also discussed at https://github.com/ropensci/rgbif/issues/326). A call to `name_usage()`, `name_lookup()`, `name_suggest()`, `name_backbone()` returns an object of class `gbif`. By assigning this class a better overview of the output is returned for `print(output)` or `output`, especially if the output is a list (`return = all`, the default value of `name_usage` for example). The args for the query are printed as well, the attributes (`attributes(output)`) has been added, same scheme used by Scott for occurrences. Only difference is the attribute `$type`: for occurrences can be `"single"` or `"many"` because the query is internally split in multiple queries. In case of `name_*()` functions only `"single"` is possible because the query is not split. We could also decide to not add this attribute for name related functions.
5. In order to realize (4) for functions `name_backbone` and `name_suggest` I changed list output to tibble dataframes. This makes the output of functions of this package consistent within the package: a list of data.frames (`return = all`, `verbose = TRUE`) of class `gbif` or a data.frame (`return = all`, `verbose = FALSE`) of class `gbif` as well.
6. Unit-tests updated/added.
7. Documentation updated and improved.

Other small things solved:

1. Get a "real" tibble df as output of `occ_search(..., return = "data")`. There was a bug while defining the class of output: the trick is to add `'tbl'` in `class(out) <- c('tbl_df', 'tbl', 'data.frame', 'gbif')` otherwise you don't get the typical nice tibble layout of df.
2. Fixing https://github.com/ropensci/rgbif/issues/328: parsing of name related errors `in name_usage()` works now for multiple taxa as well (multiple taxa are returned by using `datasetKey` input parameter: `name_usage(datasetKey = ...)`, useful for us in TrIAS).
3. Fixing https://github.com/ropensci/rgbif/issues/335. Add `issue` to the input parameters of `name_lookup()` as this is allowed by GBIF API. I did it in order to extend the applicability of function `name_issues` to output of `name_lookup()` as well (aiming to replicate the functionalities for occurrence data where `occ_issues()` can be used on outputs of functions like `occ_search()`, `occ_data()` and download related data as well). However, I noticed that GBIF API doesn't return column `issues`, so we cannot apply any filtering on such data via `name_issues()`, a bug of GBIF API. I documented it in https://github.com/gbif/gbif-api/issues/37.

### CODE REVIEW

**Comment by sckott** [comment]
thanks @damianooldoni ! Will have a look

**Comment by damianooldoni** [comment]
Running `devtoools::check()` on my laptop returns
```
0 errors | 0 warnings | 0 notes
```
It seems appveyor stops while testing functions using cassettes I created. 
```
OK: 795 SKIPPED: 29 WARNINGS: 0 FAILED: 3
  1. Error: name_issues (@test-name_issues.R#4) 
  2. Error: occ_issues (@test-occ_issues.R#4) 
  3. Error: occ_issues: occ_data type=many (@test-occ_issues.R#65)
```

Travis CI fails due to settings about secure environment variables, probably about GeoNames username?

**Comment by sckott** [comment]
thanks! having a look

File: R/gbif_issues_lookup.r [hunk 1]
```diff
@@ -0,0 +1,42 @@
+#' Lookup issue definitions and short codes
+#'
+#' @export
+#' @param issue Full name of issue, e.g, CONTINENT_COUNTRY_MISMATCH
+#' @param code An issue short code, e.g. 'ccm'
+#' @examples
+#' gbif_issues_lookup(issue = 'CONTINENT_COUNTRY_MISMATCH')
+#' gbif_issues_lookup(code = 'ccm')
+#' #' gbif_issues_lookup(issue = 'COORDINATE INVALID')
```
[the remark this answers is not in the document]
Comment (damianooldoni) [reply R/gbif_issues_lookup.r]: Done here: https://github.com/ropensci/rgbif/pull/364/commits/2d17ff7001f7f9914c99961f9e625394fbcfe153

File: tests/testthat/test-occ_search.r [hunk 2]
```diff
@@ -165,27 +174,29 @@ test_that("scientificName basic use works - no synonyms", {
     # Genus is a synonym - species rank input - species not found, so Genus rank given back
     hh <- suppressMessages(occ_search(scientificName = 'Pipistrellus hesperus', limit = 2))
   }, preserve_exact_body_bytes = TRUE)
-  
+
   expect_equal(attr(bb, "args")$scientificName, "Pulsatilla patens")
   expect_equal(bb$data$species[1], "Pulsatilla patens")
+  bb_sc_nam <- "Anemone patens subsp. multifida (Pritzel) Hult�n"
```
[the remark this answers is not in the document]
Comment (damianooldoni) [reply tests/testthat/test-occ_search.r]: Thanks! Solved here: https://github.com/ropensci/rgbif/pull/364/commits/75f853eded5a2d48bc6d441f9ba8d46dc0a3ebe9

File: R/zzz.r [hunk 3]
```diff
@@ -550,6 +550,137 @@ assert <- function (x, y) {
...
+    } else {
+      tmp <- .data
+    }
+  }
+
+  # handle downloads data
+  is_dload <- FALSE
+  if (
+    c("issue", "gbifID", "accessRights", "accrualMethod") %in% names(tmp) &&
+    !"issues" %in% names(tmp)
```
[the remark this answers is not in the document]
Comment (damianooldoni) [reply R/zzz.r]: Indeed. Thanks. Solved here:
https://github.com/ropensci/rgbif/pull/364/commits/7060eb1b3e874cacf0fc2f80a1f827a1df2c938b

I also removed `"gbifID"` in that _if_ statement as it is not only returned while downloading a DWC archive. A call to `occ_search` returns this field as well. I have just checked it by comparing the fields of this download, [0020973-190621201848488](https://www.gbif.org/occurrence/download/0020973-190621201848488), and running the code example here below:

```r
> fields_returned_by_occ_search <- names(
     occ_search(scientificName = 'Pica pica Linnaeus, 1758', 
                year = 2007,
                country = "BE",
                return = "data")
)
> "gbifID" %in% fields_returned_by_occ_search
[1] TRUE
```

[the remark this answers is not in the document]
Comment (damianooldoni) [reply R/zzz.r]: Thanks. Solved here: https://github.com/ropensci/rgbif/pull/364/commits/7060eb1b3e874cacf0fc2f80a1f827a1df2c938b

I removed `"gbifID"` from the vector in line 605 as this field is returned by `occ_search` as well.

**Comment by damianooldoni** [comment]
By tackling one of problems mentioned above (https://github.com/ropensci/rgbif/commit/c4713d35b278e2a16d43ecd6b81acb179bedbdfa) I found why Travis CI & co. didn't succeed as expected. So, this is finally solved.

About revdep compatibility, I did the following:

```r
source("https://install-github.me/r-lib/revdepcheck")
library(revdepcheck)
revdep_check(num_workers = 4)
```

I added two subfolders of `revdep` (`library` and `checks`) to `.gitignore` to avoid thousands of files being uploaded to gitHub.

No warnings or failures returned. Just `revdep/README.md` has been updated:  https://github.com/ropensci/rgbif/pull/364/commits/0ec4e2d4f6cd6e9546789a759562c14d5e6b9c99

I see a big difference in the section about dependencies between the master version:

|package |old   |new   |Δ 
|:-------|:-----|:-----|:--
|rgbif   |1.2.0 |1.3.0 |* 

and the PR version:

|package |old   |new   |<U+0394>  
|:-------|:-----|:-----|:--
|rgbif        |1.3.0    |1.3.0.9134 |*  |
|askpass      |NA       |1.1        |*  |
|assertthat   |0.2.1    |0.2.1      |   |
|backports    |NA       |1.1.4      |*  |
|BH           |1.69.0-1 |1.69.0-1   |   |
|cli          |1.1.0    |1.1.0      |   |
|colorspace   |1.4-1    |1.4-1      |   |
|crayon       |1.3.4    |1.3.4      |   |
|crul         |0.8.4    |0.8.4      |   |
|curl         |4.0      |4.0        |   |
|data.table   |1.12.2   |1.12.2     |   |
|digest       |NA       |0.6.20     |*  |
|ellipsis     |0.2.0.1  |0.2.0.1    |   |
|fansi        |0.4.0    |0.4.0      |   |
|geoaxe       |0.1.0    |0.1.0      |   |
|ggplot2      |3.2.0    |3.2.0      |   |
|glue         |NA       |1.3.1      |*  |
|gtable       |0.3.0    |0.3.0      |   |
|httpcode     |0.2.0    |0.2.0      |   |
|httr         |NA       |1.4.0      |*  |
|jsonlite     |1.6      |1.6        |   |
|labeling     |0.3      |0.3        |   |
|lazyeval     |0.2.2    |0.2.2      |   |
|mime         |NA       |0.7        |*  |
|munsell      |0.5.0    |0.5.0      |   |
|oai          |0.2.2    |0.2.2      |   |
|openssl      |NA       |1.4.1      |*  |
|pillar       |1.4.2    |1.4.2      |   |
|pkgconfig    |2.0.2    |2.0.2      |   |
|plyr         |1.8.4    |1.8.4      |   |
|R6           |NA       |2.4.0      |*  |
|RColorBrewer |1.1-2    |1.1-2      |   |
|Rcpp         |NA       |1.0.2      |*  |
|reshape2     |1.4.3    |1.4.3      |   |
|rgeos        |0.5-1    |0.5-1      |   |
|rlang        |0.4.0    |0.4.0      |   |
|scales       |1.0.0    |1.0.0      |   |
|sp           |1.3-1    |1.3-1      |   |
|stringi      |NA       |1.4.3      |*  |
|stringr      |NA       |1.4.0      |*  |
|sys          |NA       |3.2        |*  |
|tibble       |2.1.3    |2.1.3      |   |
|triebeard    |0.3.0    |0.3.0      |   |
|urltools     |1.7.3    |1.7.3      |   |
|utf8         |1.1.4    |1.1.4      |   |
|vctrs        |0.2.0    |0.2.0      |   |
|viridisLite  |0.3.0    |0.3.0      |   |
|wicket       |0.4.0    |0.4.0      |   |
|xml2         |1.2.1    |1.2.1      |   |
|zeallot      |0.1.0    |0.1.0      |   |

I don't understand why master version don't show the dependencies.

File: R/zzz.r (hunk 3 above)
[the remark this answers is not in the document]
Comment (sckott) [reply R/zzz.r]: thanks

**Comment by sckott** [comment]
thanks, all looks okay to me. Thanks for doing the revdep check. having another look

**Comment by sckott** [comment]
hmm, on running tests with this branch, I'm getting warnings with many of the tests, like `coercing argument of type 'character' to logical`

```r
options(warn=2)
res <- occ_search(
          geometry = 'POLYGON((30.1 10.1, 10 20, 20 40, 40 40, 30.1 10.1))',
          limit = 50)
bb <- res %>% occ_issues(gass84)
#> Error in all(c("issue", "accessRights", "accrualMethod")) :
#>   (converted from warning) coercing argument of type 'character' to logical
#> 
#> traceback()
#> 15: doWithOneRestart(return(expr), restart)
#> 14: withOneRestart(expr, restarts[[1L]])
#> 13: withRestarts({
#>         .Internal(.signalCondition(simpleWarning(msg, call), msg,
#>      ...
#> 12: .signalSimpleWarning("coercing argument of type 'character' to logical",
#>         base::quote(all(c("issue", "accessRights", "accrualMethod")))) at zzz.r#604
#> 11: all(c("issue", "accessRights", "accrualMethod")) %in% names(tmp) at zzz.r#604
#> 10: handle_issues(.data, is_occ = TRUE, ..., mutate = mutate) at occ_issues.r#120
#> 9: occ_issues(., gass84)
#> 8: function_list[[k]](value)
#> 7: withVisible(function_list[[k]](value))
#> 6: freduce(value, `_function_list`)
#> 5: `_fseq`(`_lhs`)
#> 4: eval(quote(`_fseq`(`_lhs`)), env, env)
#> 3: eval(quote(`_fseq`(`_lhs`)), env, env)
#> 2: withVisible(eval(quote(`_fseq`(`_lhs`)), env, env))
#> 1: res %>% occ_issues(gass84)
```

maybe in line 604 where you have `all(c("issue", "accessRights", "accrualMethod")) %in% names(tmp)` you really  meant to do: `all(c("issue", "accessRights", "accrualMethod") %in% names(tmp))` ?

**Comment by damianooldoni** [comment]
Omg! Sure. Sorry. 
Sometimes when stuff are too easy, faults occur. Just due to lack of testing. 
Solved in https://github.com/ropensci/rgbif/pull/364/commits/761754639b5af8180f77a553eba9735c19cf1710

Tests succeeded:
```r
devtools::test()
Duration: 116.3 s
== Results =====================================================================
OK:       972
Failed:   0
Warnings: 0
Skipped:  0
```

Checks succeeded as well:

```r
devtools::check()

Status: OK

R CMD check results
0 errors | 0 warnings | 0 notes
```

**Comment by sckott** [comment]
LGTM

**Pull request merged**